[Test Framework] Add documentation for testing framework (#1900)
onflow/cadence: docs/testing-framework.md

---
title: Cadence Testing Framework
---

Cadence testing framework provides a convenient way to write tests for Cadence programs in Cadence.
This functionality is provided by the Test standard-library.

Note: Test standard-library can only be used off-chain. e.g: In CLI

## Test Standard Library

Testing standard library can be imported directly into the test script.
```cadence
import Test
```

## Assertion
### assert
```cadence

> fun assert(_ condition: Bool, _ message: String)

Author: This is an implementation different: to match the rest of the functions in the `Test` stdlib. But yeah, can unify both. We can also remove this one, since it's already there in other stdlib

> fun assert(_ condition: Bool, _ message: String)

Author: Updated in https://github.com/onflow/cadence/pull/1900/commits/f1e7a85f615371f218068328be93589db5db95c7 for the docs. The behaviour is also updated in the other PR.

```
Fails a test-case if the given condition is false, and reports a message which explains how the condition is false.

The message argument is optional.

### fail
```cadence
fun fail(_ message: string)
```
Immediately fails a test-case, with a message explaining the reason to fail the test.

The message argument is optional.

### expect
The `expect` function tests a value against a matcher (see [matchers](#matchers) section), and fails the test if it's not a match.

```cadence
fun expect(_ value: AnyStruct, _ matcher: Matcher)
```

## Matchers
A matcher is an object that consists of a test function and associated utility functionality.

> ```cadence

Reviewer: Here and below: Add spaces between headings, text, and code fences.

Author: markdown automatically adds spaces, so I didn't add them manually. But no harm in adding also 👍

Author: Added in e5f3d1f03849cb7031da6e2499885f5bb42da307

        self.commitBlock()
        return results
    }

    /// Deploys a given contract, and initilizes it with the arguments.
    ///
    pub fun deployContract(
        name: String,
        code: String,
        account: Account,
        arguments: [AnyStruct]
    ): Error? {
        return self.backend.deployContract(
            name: name,
            code: code,
            account: account,
            arguments: arguments
        )
    }
}
```
The `BlockchainBackend` provides the actual functionality of the blockchain.

```cadence
/// BlockchainBackend is the interface to be implemented by the backend providers.
///
pub struct interface BlockchainBackend {

    pub fun executeScript(_ script: String, _ arguments: [AnyStruct]): ScriptResult

    pub fun createAccount(): Account

    pub fun addTransaction(_ transaction: Transaction)

    pub fun executeNextTransaction(): TransactionResult?

    pub fun commitBlock()

    pub fun deployContract(
        name: String,
        code: String,
        account: Account,
        arguments: [AnyStruct]
    ): Error?
}
```

### Creating a blockchain
A new blockchain instance can be created using the `newEmulatorBlockchain` method.

> It returns a `Blockchain` which is backed by a new emulator instance.

Reviewer: ```suggestion
It returns a `Blockchain` which is backed by a new [Flow Emulator](https://developers.flow.com/tools/emulator) instance.

```

```cadence
let blockchain = Test.newEmulatorBlockchain()
```
### Creating accounts

> It may require to create accounts during tests for various reasons such as deploying contracts, signing transactions, etc.

Reviewer: ```suggestion
It may be necessary to create accounts during tests for various reasons, such as for deploying contracts, signing transactions, etc.
```

An account can be created using the `createAccount` function.
```cadence
let acct = blockchain.createAccount()
```
The returned account consist of the `address` of the account, and a `publicKey` associated with it.
```cadence
/// Account represents info about the account created on the blockchain.
///
pub struct Account {
    pub let address: Address
    pub let publicKey: PublicKey

    init(address: Address, publicKey: PublicKey) {
        self.address = address
        self.publicKey = publicKey
    }
}
```

### Executing scripts

> Execute scripts with `executeScript` function, which returns a `ScriptResult`.

Reviewer: ```suggestion
Scripts can be run with the `executeScript` function, which returns a `ScriptResult`.
```

The result of a transaction consists of the status of the execution, and an `Error` if the transaction failed.
```cadence
/// The result of a transaction execution.
///
pub struct TransactionResult {
    pub let status: ResultStatus
    pub let error: Error?

    init(status: ResultStatus, error: Error) {
        self.status = status
        self.error = error
    }
 }
```

### Commit block
`commitBlock` block will commit the current block, and will fail if there are any un-executed transactions in the block.
```cadence
blockchain.commitBlock()
```

### Deploying contracts
A contract can be deployed using the `deployContract` function of the `Blockchain`.
```cadence
let contractCode = "pub contract Foo{ pub let msg: String;   init(_ msg: String){ self.msg = msg }   pub fun sayHello(): String { return self.msg } }"

let err = blockchain.deployContract(
    name: "Foo",
    code: contractCode,
    account: account,
    arguments: ["hello from args"],
)
```
An `Error` is returned if the contract deployment fails. Otherwise, a `nil` is returned.

### Configuring import addresses
A common pattern in Cadence projects is to define the imports as file locations and specify the addresses

> corresponding to each network in the [config](https://developers.flow.com/tools/flow-cli/configuration#contracts) file.

Reviewer: ```suggestion
corresponding to each network in the [Flow CLI configuration file](https://developers.flow.com/tools/flow-cli/configuration#contracts).
```

When writing tests for a such project, it may also require to specify the addresses to be used during the tests as well.
However, during tests, since accounts are created dynamically and the addresses are also generated dynamically,
specifying the addresses statically in a configuration file is not an option.

Hence, the test framework provides a way to specify the addresses using the
`useConfiguration(_ configs: Test.Configurations)` function in `Blockchain`.

The `Configurations` struct consists of a mapping of import locations to their addresses.
```cadence
/// Configurations to be used by the blockchain.
/// Can be used to set the address mapping.
///
pub struct Configurations {

>     pub let addresses: { String: Address }
> 
>     init(addresses: { String: Address }) {

Reviewer: ```suggestion
    pub let addresses: {String: Address}

    init(addresses: {String: Address}) {
```

pub var blockchain = Test.newEmulatorBlockchain()
pub var accounts: [Test.Account] = []

pub fun setup() {
    // Create accounts in the blockchain.

    let acct1 = blockchain.createAccount()
    accounts.append(acct1)

    let acct2 = blockchain.createAccount()
    accounts.append(acct2)

    // Set the configurations with the addresses

    blockchain.useConfiguration(Test.Configurations({
        "./contracts/FooContract": acct1.address,
        "./contracts/BarContract": acct2.address
    }))
}
```
The subsequent operations on the blockchain (e.g: contract deployment, script/transaction execution) will resolve the
file import locations to the provided addresses.

### Errors
An `Error` maybe returned when an operation (such as executing a script, executing a transaction, etc.) is failed.
Contains a message indicating why the operation failed.
```cadence
// Error is returned if something has gone wrong.
//
pub struct Error {
    pub let message: String

    init(_ message: String) {
        self.message = message
    }
}
```
An `Error` may typically be handled by failing the test case or by panicking (which will result in failing the test).
```cadence
let err: Error? = ...

> if err != nil {
>     panic(err!.message)
> }

Reviewer: ```suggestion
if let err = err {
    panic(err.message)
}
```
